File: lib/features/achievements/data/achievements_repository.dart
import 'package:clashkingapp/core/services/api_service.dart';
import 'package:clashkingapp/features/auth/data/auth_service.dart';
import 'package:clashkingapp/features/achievements/models/achievement.dart';
import 'package:flutter/foundation.dart';

class AchievementsRepository extends ChangeNotifier {
  AchievementsRepository({ApiService? apiService})
    : _apiService = apiService ?? ApiService.shared;

  final ApiService _apiService;
  List<Achievement> _achievements = const [];
  bool _isRefreshing = false;
  int _sessionGeneration = 0;
  String? _sessionUserId;
  AuthService? _authService;

  List<Achievement> get achievements => List.unmodifiable(_achievements);
  bool get isRefreshing => _isRefreshing;

  void bindAuth(AuthService authService) {
    if (identical(_authService, authService)) return;
    _authService?.removeListener(_handleAuthChanged);
    _authService = authService;
    _sessionUserId = _currentSessionUserId;
    authService.addListener(_handleAuthChanged);
  }

  String? get _currentSessionUserId => _authService?.canUseApp == true
      ? _authService?.currentUser?.userId
      : null;

  void _handleAuthChanged() {
    final nextUserId = _currentSessionUserId;
    if (nextUserId == _sessionUserId) return;
    _sessionUserId = nextUserId;
    clear();
  }

  void clear() {
    _sessionGeneration++;
    _achievements = const [];
    _isRefreshing = false;
    notifyListeners();
  }

  @override
  void dispose() {
    _authService?.removeListener(_handleAuthChanged);
    super.dispose();
  }

  Future<void> load() async {
    final generation = _sessionGeneration;
    final response = await _apiService.get('/achievements', requiresAuth: true);
    if (generation != _sessionGeneration) return;
    _replaceFromResponse(response);
  }

  Future<void> check() async {
    if (_isRefreshing) return;
    final generation = _sessionGeneration;
    _isRefreshing = true;
    notifyListeners();
    try {
      final response = await _apiService.post(
        '/achievements/check',
        const {},
        requiresAuth: true,
      );
      if (generation != _sessionGeneration) return;
      _replaceFromResponse(response);
    } finally {
      _isRefreshing = false;
      notifyListeners();
    }
  }

  void _replaceFromResponse(Map<String, dynamic> response) {
    final rawItems = response['items'];
    if (rawItems is! List) {
      throw const FormatException('Achievement response is missing items.');
    }

    final remoteById = <AchievementId, Achievement>{};
    for (final rawItem in rawItems) {
      if (rawItem is! Map) continue;
      final item = Map<String, dynamic>.from(rawItem);
      final id = AchievementIdApi.tryParse(item['id'] as String? ?? '');
      final modelUrl = item['asset_url'];
      final earnedCount = item['earned_count'];
      final repeatable = item['repeatable'];
      if (id == null ||
          modelUrl is! String ||
          modelUrl.isEmpty ||
          earnedCount is! num ||
          repeatable is! bool) {
        continue;
      }
      remoteById[id] = Achievement(
        id: id,
        modelUrl: modelUrl,
        earnedCount: earnedCount.toInt().clamp(0, 1 << 31),
        isRepeatable: repeatable,
      );
    }

    _achievements = [
      for (final catalogItem in achievementCatalogFallback)
        ?remoteById[catalogItem.id],
    ];
    notifyListeners();
  }
}

File: expo/src/features/achievements/data/achievements-repository.ts
import { ResponseFormatException, type ApiClient } from '../../../core/api/client';
import {
  ACHIEVEMENT_CATALOG_FALLBACK,
  isAchievementId,
  type Achievement,
  type AchievementId,
} from '../models';

export interface AchievementsSnapshot {
  readonly achievements: readonly Achievement[];
  readonly isRefreshing: boolean;
}

type Listener = (snapshot: AchievementsSnapshot) => void;

export class AchievementsRepository {
  private snapshotValue: AchievementsSnapshot = Object.freeze({
    achievements: Object.freeze([]),
    isRefreshing: false,
  });
  private readonly listeners = new Set<Listener>();
  private sessionGeneration = 0;
  private sessionUserId: string | null = null;

  constructor(private readonly api: Pick<ApiClient, 'requestRecord'>) {}

  get snapshot(): AchievementsSnapshot {
    return this.snapshotValue;
  }

  subscribe(listener: Listener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  bindSession(userId: string | null): void {
    if (userId === this.sessionUserId) return;
    this.sessionUserId = userId;
    this.clear();
  }

  clear(): void {
    this.sessionGeneration += 1;
    this.setSnapshot([], false);
  }

  async load(): Promise<void> {
    const generation = this.sessionGeneration;
    const response = await this.fetchCatalog();
    if (generation !== this.sessionGeneration) return;
    this.replaceFromResponse(response);
  }

  async check(): Promise<void> {
    if (this.snapshotValue.isRefreshing) return;
    const generation = this.sessionGeneration;
    this.setSnapshot(this.snapshotValue.achievements, true);
    try {
      const response = await this.fetchCatalog();
      if (generation !== this.sessionGeneration) return;
      this.replaceFromResponse(response, true);
    } finally {
      if (generation === this.sessionGeneration) {
        this.setSnapshot(this.snapshotValue.achievements, false);
      }
    }
  }

  private fetchCatalog(): Promise<Record<string, unknown>> {
    return this.api.requestRecord('/achievements/check', {
      method: 'POST',
      body: {},
      requiresAuth: true,
    });
  }

  private replaceFromResponse(response: Record<string, unknown>, preserveRefreshing = false): void {
    if (!Array.isArray(response.items)) {
      throw new ResponseFormatException('Achievement response is missing items.');
    }
    const remoteById = new Map<AchievementId, Achievement>();
    for (const rawItem of response.items) {
      if (!isRecord(rawItem)) continue;
      const id = typeof rawItem.id === 'string' ? rawItem.id : '';
      const modelUrl = rawItem.asset_url;
      const earnedCount = rawItem.earned_count;
      const repeatable = rawItem.repeatable;
      if (
        !isAchievementId(id) ||
        typeof modelUrl !== 'string' ||
        modelUrl.length === 0 ||
        typeof earnedCount !== 'number' ||
        !Number.isFinite(earnedCount) ||
        typeof repeatable !== 'boolean'
      ) {
        continue;
      }
      remoteById.set(id, {
        id,
        modelUrl,
        earnedCount: Math.min(2 ** 31, Math.max(0, Math.trunc(earnedCount))),
        isRepeatable: repeatable,
      });
    }
    const achievements = ACHIEVEMENT_CATALOG_FALLBACK.flatMap((catalogItem) => {
      const remote = remoteById.get(catalogItem.id);
      return remote ? [remote] : [];
    });
    this.setSnapshot(achievements, preserveRefreshing && this.snapshotValue.isRefreshing);
  }

  private setSnapshot(achievements: readonly Achievement[], isRefreshing: boolean): void {
    this.snapshotValue = Object.freeze({
      achievements: Object.freeze([...achievements]),
      isRefreshing,
    });
    this.listeners.forEach((listener) => listener(this.snapshotValue));
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
